plat/kvm/x86/multiboot2.c

```diff
@@ -0,0 +1,208 @@
+
+static inline void mrd_insert(struct ukplat_bootinfo *bi,
+			      const struct ukplat_memregion_desc *mrd)
+{
+	int rc;
+
+	if (unlikely(mrd->len == 0))
+		return;
+
+	rc = ukplat_memregion_list_insert(&bi->mrds, mrd);
+	if (unlikely(rc < 0))
+		multiboot2_crash("Cannot insert bootinfo memory region", rc);
+}
+
+static void multiboot2_init_mem(struct ukplat_bootinfo *bi,
+				struct multiboot_tag *tag)
+{
+	struct multiboot_tag_mmap *mmap_tag = (struct multiboot_tag_mmap *)tag;
+
+	struct multiboot_mmap_entry *mmap = mmap_tag->entries;
+	struct ukplat_memregion_desc mrd = {0};
+	__paddr_t start, end;
+	__sz offset;
+
+	__sz mmap_entry_cnt = (mmap_tag->size - sizeof(struct multiboot_tag_mmap)) /
+			    		 mmap_tag->entry_size;
+
+	for (offset = 0; offset < mmap_entry_cnt;
+		offset++) {
+		multiboot_memory_map_t *m = mmap + offset;
+
+		start = MAX(m->addr, __PAGE_SIZE);
+		end   = m->addr + m->len;
+		if (unlikely(end <= start || end - start < PAGE_SIZE))
+			continue;
+
+		mrd.pbase = PAGE_ALIGN_DOWN(start);
+		mrd.vbase = mrd.pbase; /* 1:1 mapping */
+		mrd.pg_off = start - mrd.pbase;
+		mrd.len = end - start;
+		mrd.pg_count = PAGE_COUNT(mrd.pg_off + mrd.len);
+
+		if (m->type == MULTIBOOT_MEMORY_AVAILABLE) {
+			mrd.type  = UKPLAT_MEMRT_FREE;
+			mrd.flags = UKPLAT_MEMRF_READ |
+					UKPLAT_MEMRF_WRITE;
+
+			/* Free memory regions have
+			* mrd.len == mrd.pg_count * PAGE_SIZE
+			*/
+			mrd.len = PAGE_ALIGN_UP(mrd.len + mrd.pg_off);
+		} else {
+			mrd.type  = UKPLAT_MEMRT_RESERVED;
+			mrd.flags = UKPLAT_MEMRF_READ;
+
+			/* We assume that reserved regions cannot
+			* overlap with loaded modules.
+			*/
+		}
+
+		mrd_insert(bi, &mrd);
+	}
+}
+
+
+/**
+ * Multiboot entry point called after lcpu initialization. We enter with the
+ * 1:1 boot page table set. Physical and virtual addresses thus match for all
+ * regions in the mapped range.
+ */
+void multiboot2_entry(struct lcpu *lcpu, unsigned long addr)
+{
+	struct ukplat_bootinfo *bi;
+	struct ukplat_memregion_desc mrd = {0};
+	multiboot_memory_map_t *m;
+	__u64 cmdline_str = {0};
+	__sz i, cmdline_len, mmap_entry_cnt;
+	__paddr_t start, end;
+	int rc;
+
+	bi = ukplat_bootinfo_get();
+	if (unlikely(!bi))
+		multiboot2_crash("Incompatible or corrupted bootinfo", -EINVAL);
+
+	/* We have to call this here as the very early do_uk_reloc32 relocator
+	 * does not also relocate the UKPLAT_MEMRT_KERNEL mrd's like its C
+	 * equivalent, do_uk_reloc, does.
+	 */
+	do_uk_reloc_kmrds(0, 0);
+
+	/* Ensure that the memory map contains the legacy high mem area */
+	rc = ukplat_memregion_list_insert_legacy_hi_mem(&bi->mrds);
+	if (unlikely(rc))
+		multiboot2_crash("Could not insert legacy memory region", rc);
+
+	/* Parse Multiboot2 tags */
+	for (struct multiboot_tag *tag = (struct multiboot_tag *)(addr + 8);
+	     tag->type != MULTIBOOT_TAG_TYPE_END;
+	     tag = (struct multiboot_tag *)((multiboot_uint8_t *)tag
+					    + ((tag->size + 7) & ~7))) {
```

Comment:
I will want some comment or macro surrounding this as it does not tell me much about this formula.